real_preprocessing/src/corner_detection.cpp

```diff
@@ -0,0 +1,88 @@
+#include "ros/ros.h"
+#include "apriltag_ros/AprilTagDetectionArray.h"
+#include "sstream"
+#include "tf_conversions/tf_eigen.h"
+#include "Eigen/Dense"
+#include "fstream"
+#include "string"
+
+void tfCallback(const apriltag_ros::AprilTagDetectionArray::ConstPtr& msg)
+{
+  Eigen::MatrixXd intrinsic(3, 3);  // potential calibration function
+  intrinsic << 630.3467672874693, 0, 314.10832937485145, 0, 624.9044125522948, 241.59156511756711, 0, 0, 1;
+  if (msg->detections.empty())
+  {
+    // no detections
+  }
+  else
+  {
+    std::string filename("detections_0000.yml");  // output path
+    std::ofstream fout;
+    fout.open(filename.c_str());
+    fout << "detections:";
+
+    for (float i = (msg->detections.size() - 1); i > -1; i--)
+    {
+      // reading data from topic
+      float xpos = msg->detections[i].pose.pose.pose.position.x;
+      float ypos = msg->detections[i].pose.pose.pose.position.y;
+      float zpos = msg->detections[i].pose.pose.pose.position.z;
+      float xori = msg->detections[i].pose.pose.pose.orientation.x;
+      float yori = msg->detections[i].pose.pose.pose.orientation.y;
+      float zori = msg->detections[i].pose.pose.pose.orientation.z;
+      float wori = msg->detections[i].pose.pose.pose.orientation.w;
+      float size = msg->detections[i].size[0];
+      int id = msg->detections[i].id[0];
+
+      tf::Quaternion q(xori, yori, zori, wori);
+      tf::Matrix3x3 m(q);  // quaternion to rotational matrix
+
+      Eigen::MatrixXd trans(3, 4);  // 3x4 transformation matrix
+      trans << m[0][0], m[0][1], m[0][2], xpos, m[1][0], m[1][1], m[1][2], ypos, m[2][0], m[2][1], m[2][2], zpos;
```

Comment:
The variable names are not very descriptive. We generally try to include the parent and child frames in the variable name for example, I would call this cam_T_tag.